htlcswitch/switch.go

```diff
@@ -892,14 +874,37 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 			return s.failAddPacket(packet, failure, addErr)
 		}
 		interfaceLinks, _ := s.getLinks(targetLink.Peer().PubKey())
+		s.indexMtx.Unlock()
+
+		// We'll keep track of any HTLC failures during the link
+		// selection process. This way we can return the error for
+		// precise link that the sender selected, while optimistically
+		// trying all inks to utilize out available bandwidth.
+		linkErrs := make(map[lnwire.ShortChannelID]lnwire.FailureMessage)
 
 		// Try to find destination channel link with appropriate
 		// bandwidth.
 		var destination ChannelLink
 		for _, link := range interfaceLinks {
 			// We'll skip any links that aren't yet eligible for
 			// forwarding.
-			if !link.EligibleToForward() {
+			switch {
+			case !link.EligibleToForward():
+				fallthrough
+
+			case link.ShortChanID().ToUint64() == 0:
```

Comment:
Should add comment to explain that this should never happen, but protects against invalid circuit map states, i.e. that we should never be forwarding an HTLC _out_ to the source hop.
Maybe replace with `case link.ShortChanID() == sourceHop:` to match comment?